### ISSUE TITLE
Passing parameters made by get_parameter() to submodules

### ISSUE THREAD

**Issue by Pascal314** [deepmind/dm-haiku#372]
If I want to reparameterize existing submodules, I can use get_parameter() to define the parameterization that I want and rewrite the functionality of the submodule. For example, here is a minimal version of Conv2D with a rank 1 filter parameterized in its singular value decomposition.
```
def forward(x):
    u = hk.get_parameter('u', (n, 1), init=hk.initializers.RandomNormal())
    v = hk.get_parameter('v', (1, n), init=hk.initializers.RandomNormal())
    w = np.matmul(u, v)


    y = jax.lax.conv(jnp.transpose(x, (0, 3, 1, 2)), 
                     jnp.broadcast_to(w, (1, 1, n, n)) , 
                     window_strides=(1, 1), 
                     padding='same')
                
    return jnp.transpose(y, (0, 2, 3, 1))
```
I would rather avoid writing the convolution myself, and instead pass `w` to the haiku `Conv2D` implementation.
Can I somehow pass `w` to the `Conv2D` module so that I don't have to worry about implementing the convolution/broadcasting/tests?

I would imagine this would look something like this:

```
def forward(x):
    u = hk.get_parameter('u', (n, 1), init=hk.initializers.RandomNormal())
    v = hk.get_parameter('v', (1, n), init=hk.initializers.RandomNormal())
    w = np.matmul(u, v)

    y = hk.Conv2D(1, (n, n), use_parameter=w)
    return jnp.transpose(y, (0, 2, 3, 1))
```
If there is a clever way to do this I would like to write a wrapper that takes in a `hk.Module` and implements weight normalization: https://arxiv.org/abs/1602.07868

**Comment by Pascal314** [comment]
I realized that it is possible to simply transform the module first:
```
def forward(x):
    u = hk.get_parameter('u', (n, 1), init=hk.initializers.RandomNormal())
    v = hk.get_parameter('v', (1, n), init=hk.initializers.RandomNormal())
    w = np.matmul(u, v)

    def conv(x):
        return hk.Conv2D(1, (n, n), with_bias=False)(x)
    
    params = {'conv2_d': {'w': jnp.reshape(w, (n, n, 1, 1))}}
    conv = hk.without_apply_rng(hk.transform(conv))
    y = conv.apply(params, x)

    return y
```
Is there a method that allows me to infer the tree structure and names such as 'conv2_d' and 'w' so that I don't have to guess them, ideally without calling init on the inner module?
Is there a good reason not to do this?

**Comment by tomhennigan** [comment]
Hi @Pascal314 , we have [`hk.custom_getter`](https://dm-haiku.readthedocs.io/en/latest/api.html#haiku.custom_getter) and [`hk.custom_creator`](https://dm-haiku.readthedocs.io/en/latest/api.html#haiku.custom_creator) which allow you to intercept retrieval and creation of parameters for any module. You could use this to change the way parameters are created or mutated for any module (including built in ones).
